src/libopensc/card-jpki.c

```diff
@@ -0,0 +1,353 @@
+	       sc_print_path(path), path->len);
+	if (path->len == 2 && memcmp(path->value, "\x3F\x00", 2) == 0) {
+		drvdata->selected = SELECT_MF;
+		if (file_out) {
+			sc_file_dup(file_out, drvdata->mf);
+			if (*file_out == NULL) {
+				LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+			}
+		}
+		return 0;
+	}
+
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0xA4, 0, 0);
+	switch (path->type) {
+	case SC_PATH_TYPE_FILE_ID:
+		apdu.p1 = 2;
+		break;
+	case SC_PATH_TYPE_DF_NAME:
+		apdu.p1 = 4;
+		break;
+	default:
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+	}
+	apdu.p2 = 0x0C;
+	apdu.data = path->value;
+	apdu.datalen = path->len;
+	apdu.lc = path->len;
+
+	rc = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(card->ctx, rc, "APDU transmit failed");
+	rc = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	LOG_TEST_RET(card->ctx, rc, "SW Check failed");
+	if (file_out) {
+		file = sc_file_new();
+		if (!file) {
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+		}
+		file->path = *path;
+		file->size = 2048;
+		*file_out = file;
+	}
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int
+jpki_read_binary(sc_card_t * card, unsigned int idx,
+		 u8 * buf, size_t count, unsigned long flags)
+{
+	struct sc_card_driver *iso_drv = sc_get_iso7816_driver();
+	const struct sc_card_operations *iso_ops = iso_drv->ops;
+	int rc;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	rc = iso_ops->read_binary(card, idx, buf, count, flags);
+	LOG_FUNC_RETURN(card->ctx, rc);
+}
+
+static int
+jpki_pin_cmd(sc_card_t * card, struct sc_pin_cmd_data *data, int *tries_left)
+{
+	int rc;
+	sc_path_t path;
+	sc_apdu_t apdu;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (tries_left) {
+		*tries_left = -1;
+	}
+
+	switch (data->pin_reference) {
+	case 1:
+		sc_format_path(JPKI_AUTH_PIN, &path);
+		path.type = SC_PATH_TYPE_FILE_ID;
+		rc = sc_select_file(card, &path, NULL);
+		break;
+	case 2:
+		sc_format_path(JPKI_SIGN_PIN, &path);
+		path.type = SC_PATH_TYPE_FILE_ID;
+		rc = sc_select_file(card, &path, NULL);
+		break;
+	default:
+		sc_log(card->ctx, "Unknown PIN reference: %d", data->pin_reference);
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+	}
+	LOG_TEST_RET(card->ctx, rc, "SELECT_FILE error");
+
+	switch (data->cmd) {
+	case SC_PIN_CMD_VERIFY:
+		sc_format_apdu(card, &apdu, SC_APDU_CASE_3, 0x20, 0x00, 0x80);
+		apdu.data = data->pin1.data;
+		apdu.datalen = data->pin1.len;
+		apdu.lc = data->pin1.len;
+		rc = sc_transmit_apdu(card, &apdu);
+		LOG_TEST_RET(card->ctx, rc, "APDU transmit failed");
+		rc = sc_check_sw(card, apdu.sw1, apdu.sw2);
+		LOG_TEST_RET(card->ctx, rc, "VERIFY failed");
+		break;
+	case SC_PIN_CMD_GET_INFO:
+		sc_format_apdu(card, &apdu, SC_APDU_CASE_1, 0x20, 0x00, 0x80);
+		rc = sc_transmit_apdu(card, &apdu);
+		LOG_TEST_RET(card->ctx, rc, "APDU transmit failed");
+		if (apdu.sw1 != 0x63) {
+			sc_log(card->ctx, "VERIFY GET_INFO error");
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_CARD_CMD_FAILED);
```

Comment:
If your card is ISO 7816 compliant, then you should get `9000` when the user is logged with this APDU. In this case you should not return an error!